### ISSUE TITLE
ActiveRecord: PostgreSQL hstore null character terminates value insert string

### ISSUE THREAD

**Issue by daveallie** [rails/rails#34248]
### Steps to reproduce

- Create a PostgreSQL ActiveRecord model with a hstore column
- Attempt to save a hstore value with a null character in it
```ruby 
record = TestRecord.new(hstorecol: {a: "te\0st"})
puts record.attributes
#=> {"id"=>nil, "hstorecol"=>{"a"=>"te\u0000st"}

record.save!
#   (24.9ms)  BEGIN /*line:(irb):12:in `irb_binding',hostname:daveallie,pid:579*/
#  SQL (25.8ms)  INSERT INTO "test_records" ("hstorecol") VALUES ('"a"=>"te') RETURNING "id" /*line:(irb):12:in `irb_binding',hostname:daveallie,pid:579*/
#   (19.5ms)  ROLLBACK /*line:(irb):12:in `irb_binding',hostname:daveallie,pid:579*/
# ActiveRecord::StatementInvalid: PG::InternalError: ERROR:  Unexpected end of string
# LINE 1: INSERT INTO "test_records" ("hstorecol") VALUES ('"a"=>"te')...
#                                                          ^
# : INSERT INTO "test_records" ("hstorecol") VALUES ('"a"=>"te') RETURNING "id" /*line:(irb):12:in `irb_binding',hostname:daveallie,pid:579*/
#	from (irb):12
```

### Expected behavior
Null character should be escaped or ignored. Should store something like: `'"a"=>"te\\u0000st"'` or just strip null characters, storing `'"a"=>"test"'`.

### Actual behavior
Null character terminates column string resulting in an invalid hstore. This is a pontential SQL injection vulnerability? Is this an issue with hstore serialization code here: https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/postgresql/oid/hstore.rb#L26..L34

### System configuration
**Rails version**: 5.1.6

**Ruby version**: 2.3.3

**Comment by y-yagi** [comment]
I tried your issue, but not reproduce. Attempt to save an hstore value with a null character raises `ArgumentError`.  

```ruby  
#                                      Table "public.foos"
#    Column   |            Type             |                     Modifiers                     
# ------------+-----------------------------+---------------------------------------------------
#  id         | bigint                      | not null default nextval('foos_id_seq'::regclass)
#  store      | hstore                      | 
#  created_at | timestamp without time zone | not null
#  updated_at | timestamp without time zone | not null 

class Foo < ApplicationRecord
end 
``` 

```ruby 
Foo.create!(store: {a: "te\0st"})
  Foo Create (5.6ms)  INSERT INTO "foos" ("store", "created_at", "updated_at") VALUES ($1, $2, $3) RETURNING "id"  [["store", "\"a\"=>\"te\u0000st\""], ["created_at", "2018-10-18 23:44:07.903492"], ["updated_at", "2018-10-18 23:44:07.903492"]]
   (0.2ms)  ROLLBACK
Traceback (most recent call last):
        1: from (irb):1
ArgumentError (string contains null byte)
``` 

Can you create a reproduce script with this [template](https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_gem.rb)

**Comment by daveallie** [comment]
Apologies, this was an issue with the `pg` gem and not using prepared statements, resolved in v1.1, we are not on the 1.0 release yet for `pg` and I wrongly assumed it was an AR issue.

Here is the replication:

Changing the `pg` gem version to `1.1` resolves the error.

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "5.2.1"
  gem "pg", "0.21"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with minitest 4.
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "rails_testing", username: `whoami`.strip, prepared_statements: false)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.hstore :store
  end
end

class Post < ActiveRecord::Base; end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!
    post.store = {a: "te\0st"}
    post.save
  end
end
```
